### ISSUE TITLE
Add TriBITS support for all-at-once configure, build, test, and submit to CDash

### ISSUE THREAD

**Issue by bartlettroscoe** [TriBITSPub/TriBITS#183]
**Description:**

This Story is to track TriBITS changes needed to take advantage of the new CMake/CTest and CDash feature being created to so a single configure, build, test, and submit to CDash for mutiple subprojects (i.e. packages) being tracked in:

* https://gitlab.kitware.com/snl/project-1/issues/8

**Tasks:**

1. Add support for TribitsCTestDriverCore.cmake to TribitsExampleProject and add some automated tests for some core use cases (no git operations and no email) (We need to cover the code before we change it) ... See #193 **[Done]**

2. Refactor TribitsCTestDriverCore.cmake to allow for a single configure, build, and submit. **[Done]**

3. Do trial submit for Trilinos packages to site http://exp.cdash.org/index.php?project=TribitsExProj. **[Done]**

4. Provide instructions to Kitware for accessing this branch and pushing development of this feature. **[Done]**

5. Wait for Kitware to complete changes to CMake/CTest and CDash and adjust TriBITS as needed ... **IN PROGRESS ...**

**Comment by bartlettroscoe** [comment]
Example/test code in CMake for how to take advantage of this is pointed out in:

* https://gitlab.kitware.com/snl/project-1/issues/8#note_254115

From that, it looks like the command `set_directory_properties(PROPERTIES LABELS "${PACKAGE_NAME}")` followed by the command `ctest_add_subproject(${PACKAGE_NAME} )` needs to be called for each TriBITS package (a.k.a. CMake/CDash Subproject).  The question is when these commands need to be called in relation to the definition of the various targets and tests for a given TriBITS package.  From what I see in the example code, it seems that calling `set_directory_properties(PROPERTIES LABELS "${PACKAGE_NAME}")`  in `TRIBITS_PACKAGE_DECL()` and calling `ctest_add_subproject(${PACKAGE_NAME} )` in `TRIBITS_PACKAGE_POSTPROCESS()` might do the trick (at least that is what I can try).

But there is some defect in the corresponding CDash branch that need to be fixed before we can really review what this looks like on CDash.

**Comment by ibaned** [comment]
Here are some CMake functions I wrote to do something similar (config, build, test all posting to CDash) in case they can be of any help as a reference:
https://github.com/gahansen/Albany/blob/master/doc/LCM/build/snl_helpers.cmake

**Comment by bartlettroscoe** [comment]
@ibaned:

> Here are some CMake functions I wrote to do something similar (config, build, test all posting to CDash) in case they can be of any help as a reference:
> https://github.com/gahansen/Albany/blob/master/doc/LCM/build/snl_helpers.cmake

Thanks for the pointer.

The issue is that changes are required to CMake/CTest and CDash to really do what we want.  Those changes are in branch by Kitware we just need to refactor TriBITS to take advantage of that.  That is what I need to do ASAP.

**Comment by bartlettroscoe** [comment]
After a lot of work, I have pushed changes to TriBITS in this github 'master' branch and snapshotted to Trilinos that allow for an all-at-once configure, build, test and submit.  My detailed notes are given below.

From the updated [tribits/ReleaseNotes.txt](https://github.com/TriBITSPub/TriBITS/blob/3843b1a2a5fba9dddd41aeba82b4772a0e20e758/tribits/ReleaseNotes.txt) file:

```
2017/06/24:

(*) MINOR: Add new all-at-once more for CTest -S driver scripts using
    TRIBITS_CTEST_DRIVER() by setting the variable
    ${PROJECT_NAME}_CTEST_DO_ALL_AT_ONCE=TRUE.  This works with older versions
    of CMake/CTest and CDash but, by default, will just return a single glob
    of results, not breaking results out package-by-package.  Therefore, this
    is disabled by default and package-by-package mode is used by default.
    But if ${PROJECT_NAME}_CTEST_USE_NEW_AAO_FEATURES=TRUE is set, then
    TriBITS will take advantage of new CMake, CTest, and CDash features
    (currently on a branch) to display the results on CDash broken down
    package-by-package.  Once these changes are merged to the CMake/CTest and
    CDash 'master' branches, then the default for
    ${PROJECT_NAME}_CTEST_USE_NEW_AAO_FEATURES will be set to TRUE
    automatically when it detects an updated version of CMake/CTest is
    present.  In the future, at some point, the TriBITS default for
    ${PROJECT_NAME}_CTEST_DO_ALL_AT_ONCE will change from FALSE to TRUE since
    that is a much more efficient way to drive automated testing.
```


**DETAILED NOTES:**

(6/12/2017)

**3) Refactor TribitsCTestDriverCore.cmake to allow for a single configure, build, and submit to exp.cdash.org and testing.sandia.gov for only all-passed use case:**

**3.a) Break out package-by-package processing as its own macro and pass automated tests and trial submit:**

Already I am seeing an issue with this refactoring.  The issue is that the functioning of the code is dependent on knowing what packages have failures in the variable `${PROJECT_NAME}_FAILED_PACKAGES`.  That is used to write a file `failedPackages.tx` that is read on the next CI iteration to make sure and enable packages that were previously failing (even if they have not been changed).  This is also used to determine if the script returns 0 or non-zero (see https://github.com/TriBITSPub/TriBITS/issues/154).

Anyway, I was able to fator out the macro TRIBITS_CTEST_PACKAGE_BY_PACKAGE() and all the automated tests pass.

(6/13/2017)

**3.b) Add new macro that processes all packages at once (should be much simpler):**

* Create new variable `${PROJECT_NAME}_CTEST_DO_ALL_AT_ONCE` and define new seleton macro `TRIBITS_CTEST_ALL_AT_ONCE()` [Done]

* Create comments for existing package-by-package testing and new all-at-once testing [Done]

* Create new failing test for all-at-once all passing and to submit to CDash [Done]

* Fill in new macro `TRIBITS_CTEST_ALL_AT_ONCE()` until new test passes [Done]

* Submit first AAO test case to exp.cdash.org [Done]

(6/14/2017)

**4) Take advantage of new CTest/CDash features to submit all at once (AAO) appraoch and submit to exp.cdash.org and testing.sandia.gov for all passing case:**

* Add calls to CTEST_ADD_SUBPROJECT() [Done]

* Change names of package-by-package tests to _PBP_ [Done]

* Add changes to `make dashboard` so that it does the all-at-once approach if `${PROJECT_NAME}_CTEST_DO_ALL_AT_ONCE` and	`${PROJECT_NAME}_USE_NEW_AAO_CTEST_CDASH_FEATURES` are set [Done]

* Clean up commits some and push updated branch [Done]

**5) Do initial review of all-at-once CMake/CTest/CDash features.

See:

* https://gitlab.kitware.com/snl/project-1/issues/8#note_279530

[Done]

(6/15/2017)

**5) Add functionality and tests for various failing use cases AAO appraoch:**

**5.a)** Change CONFIGURE_OPTIONS to just pass in   `${PROJECT_NAME}_ENABLE_ALL_PACKAGES` when that is set and then add explicit disables for the inner configuration.  (That will cut down on the length of the configure line by a lot for that common case.):

[Done]

**5.b)** See if SET_DIRECTORY_PROPERTIES() is in older versions of CMake or not (so we can always set it or not):


No, it looks like it was added on the 'labels_to_subprojects' branch in the commit:

```
commit 36a3158158d5398bb3fa34b514227e5a299147e5
Author: Betsy McPhail <betsy.mcphail@kitware.com>
Date:   Mon Jun 12 14:07:31 2017 -0400

    Add 'LABELS' as a directory property
    
    The specified LABELS will be passed down to subdirectories as well as
    any targets or tests in the directory.

M       Source/CTest/cmCTestTestHandler.cxx
M       Source/cmLocalGenerator.cxx
```

which is *NOT* on the 'master' branch of CMake.  So that has to be protected.

[Done]

**5.c)** Create a macro that create tests for that AAO appraoch that either uses `TribitsExProj_USE_NEW_AAO_CTEST_CDASH_FEATURES=ON` or not.  Call the latter tests "AAOP" so we can do submits of both versions at the same time to compare.  (We may need to use the all-at-once appraoch before we can upgrade CDash so we need to test what we can do with current CMake/CTest/CDash.)

I did that and submitted the tests:

```
1/4 Test #252: TriBITS_CTestDriver_AAOP_ST_PackagesSubset_PASS ...   Passed   17.53 sec
2/4 Test #250: TriBITS_CTestDriver_AAO_ST_PackagesSubset_PASS ....   Passed   19.33 sec
3/4 Test #249: TriBITS_CTestDriver_AAO_ST_ALL_PASS ...............   Passed   19.84 sec
4/4 Test #251: TriBITS_CTestDriver_AAOP_ST_ALL_PASS ..............   Passed   21.54 sec

100% tests passed, 0 tests failed out of 4

Label Time Summary:
TriBITS    =  78.24 sec (4 tests)

Total Test time (real) =  21.55 sec
```

This submitted to:

* http://exp.cdash.org/index.php?project=TribitsExProj&date=2017-06-15

From that outer view, these all-at-once builds don't look that different.  They both display the same number of tests.  They both allow you to directly click on the number of configure errors and warnings, build errors and warnings, and the number of non run, fail, and pass tests.

The difference is what happens when you click on the build name.

A few interesting things.  First, for the new TriBITS_CTestDriver_AAOP_ST_PackagesSubset_PASS test, the package-by-package build summary:

* http://exp.cdash.org/index.php?project=TribitsExProj&parentid=816

only shows the two packages specifically tested.  That is good (because it
keeps the dashboard nice and titly but what happens if there are build errors
or warnings in the other packages?  I will have to add a test for that and see
what happens.  I added a todo for that.

[Done]

**5.d)** Try to submit AAOP tests to old testing.sandia.gov TribitsExProj project and see what happens:

To test this I did:

```
$ rm -r CMake*
$ ./do-configure -DTriBITS_TRIBITSEXPROJ_SUBMIT_TO=TRILINOS_CDASH
$ ctest -j16 -R CTestDriver_AAO
Test project /ascldap/users/rabartl/Trilinos.base/BUILDS/GCC-4.8.3/TRIBITS_MPI_DEBUG
    Start 251: TriBITS_CTestDriver_AAOP_ST_ALL_PASS
    Start 252: TriBITS_CTestDriver_AAOP_ST_PackagesSubset_PASS
    Start 250: TriBITS_CTestDriver_AAO_ST_PackagesSubset_PASS
    Start 249: TriBITS_CTestDriver_AAO_ST_ALL_PASS
1/4 Test #250: TriBITS_CTestDriver_AAO_ST_PackagesSubset_PASS ....   Passed   11.19 sec
2/4 Test #252: TriBITS_CTestDriver_AAOP_ST_PackagesSubset_PASS ...   Passed   11.49 sec
3/4 Test #251: TriBITS_CTestDriver_AAOP_ST_ALL_PASS ..............   Passed   13.10 sec
4/4 Test #249: TriBITS_CTestDriver_AAO_ST_ALL_PASS ...............   Passed   13.12 sec

100% tests passed, 0 tests failed out of 4

Label Time Summary:
TriBITS    =  48.89 sec (4 tests)

Total Test time (real) =  13.14 sec
```

This submitted those tests to:

* https://testing.sandia.gov/cdash/index.php?project=TribitsExProj&date=2017-06-15

All of the results are showing up for both types of build.  The "_AAOP_" builds actually don't look that bad when submitted to the old CDash site.  One strange thing is the the "Labels" column shows the last subproject name "WrapExternal".  For the "_AAO_" builds, it shows "(none)".


The build `TriBITS_CTestDriver_AAOP_ST_ALL_PASS` shown at:

* https://testing.sandia.gov/cdash/index.php?project=TribitsExProj&date=2017-06-15

shows a "Subprojects" column with a single value "(4 labels)".  Bit it lets you directly click on each of the numbers and see all of the tests.  That is not terrible.  In some ways, it is actually better than the `TriBITS_CTestDriver_AAO_ST_ALL_PASS` build shown at:

* https://testing.sandia.gov/cdash/buildSummary.php?buildid=2949867

which is just the flat build page.

Therefore, this means that we could do the all-at-once build with an updated version of CMake/CTest with the new all-at-once features and then do a dual a dual submit to both a new CDash site and an old CDash site.  Therefore, I think I could do dual submits for the Trilinos CI server right away.  This is very good news.

[Done]

(6/16/2017)

**5.e)** Add  failing configure cases that can submit to exp.cdash.org correctly.

The first test to add is a failed configure.  A failed configure should trigger the enable of all packages on the next CI iteration.  For that purpose, we can write "ALL_PACKAGES" to the failedPackages.txt file if `${PROJECT_NAME}_ENABLE_ALL_PACKAGES` was set.  That will then need to be interpreted correctly on the next CI iteration.  That will require some updates to the logic in the TribitsCTestDriver.cmake code (see later cleanup).  The other use case is when a subset of packages is tested.  In that case, it should just list the subset of packages ... I have the configure tests passing and skipping the build and testing steps (there is no point if the configure fails).

Next I need to test failures to build libraries and tests.  For that, I need the detection of failing tests and matching them up to the names of top-level TriBITS packages.

(6/19/2017)

**5.f)** Add functionality to detect failing packages by examining `LastTestsFailed*.log` file (write a Python scrpt that reads `<Project>PackageDependencies.cmake` and `LastTestsFailed*.log` and returns the list of failed packages as a comma-separated string). (Write a bunch of Python unit tests to test various use cases):

I think this will not be too hard.  I can just base this on the existing code that maps from a directory path to a package name.  To compare, here is how TribitsCTestCoreBase.cmake maps from directory paths to package names ...

* The macro `ENABLE_ONLY_MODIFIED_PACKAGES()` called from `TRIBITS_CTEST_DRIVER()` executes the command `get-tribits-packages-from-files-list.py --files-list-file=${MODIFIED_FILES_FILE_NAME}       --deps-xml-file=${CTEST_BINARY_DIRECTORY}/${${PROJECT_NAME}_PACKAGE_DEPS_XML_FILE_NAME}`.  It returns the list of modified packages from the `OUTPUT_VARIABLE`.

* The script `get-tribits-packages-from-files-list.py` calls `TribitsPackageFilePathUtils.getPackagesListFromFilePathsList()`.

* `TribitsPackageFilePathUtils.getPackagesListFromFilePathsList()` for each file path calls `TribitsDependencies.getPackageNameFromPath()`.

* `TribitsDependencies.getPackageNameFromPath()` loops through all the packages one at a time and uses a regex to match the file path.

(SIDENOTE: Darn, that is an expensive algorithm!  But I don't know how to make that cheaper given the generic nature of file paths.)

Looks like I can parallel set of function cals for getting the name of a package given the name of a test.  I will also copy the unit tests from the bottom up.

Got it! [Done]

**5.g)** Update AAO mode for TribitsCTestDriverCore.cmake that uses detection of failed packages based on `LastTestsFailed*.log` (add first automated test that uses this):

I have added the first passing tests with the full feature:

```
$ make && ctest -j16 -R ^TriBITS_CTestDriver_AAO.*_ST_PackagesSubset_BreakWithSubpackagesALib$
[100%] Built target TriBITS_GenericDriver
Test project /ascldap/users/rabartl/Trilinos.base/BUILDS/GCC-4.8.3/TRIBITS_MPI_DEBUG
    Start 255: TriBITS_CTestDriver_AAO_ST_PackagesSubset_BreakWithSubpackagesALib
    Start 260: TriBITS_CTestDriver_AAOP_ST_PackagesSubset_BreakWithSubpackagesALib
1/2 Test #260: TriBITS_CTestDriver_AAOP_ST_PackagesSubset_BreakWithSubpackagesALib ...   Passed   14.92 sec
2/2 Test #255: TriBITS_CTestDriver_AAO_ST_PackagesSubset_BreakWithSubpackagesALib ....   Passed   15.36 sec

100% tests passed, 0 tests failed out of 2

Label Time Summary:
TriBITS    =  30.28 sec (2 tests)

Total Test time (real) =  15.37 sec
```

I set that up to submit to exp.cdash.org and it produced:

* http://exp.cdash.org/buildSummary.php?buildid=841
* http://exp.cdash.org/index.php?project=TribitsExProj&parentid=842

The AAO build showed the build failure for the A.cpp file and the lib libpws_a at:

* http://exp.cdash.org/viewBuildError.php?buildid=841

What is interested is that the AAOP build also showed build failures for A.cpp and libpws_a.  I am not sure how CTest/CDash was able to link this up but it did.

Anyway, so far so goed.

[Done]


**5.h)** Add various failing build, and test cases that can submit to exp.cdash.org correctly that detect failing packages from matching the names of failing tests to the names of packages:

I added the tests:

```
$ make && ctest -j16 -R ^TriBITS_CTestDriver_AAO.*_ST_ALL_BreakWithSubpackagesALib$
[100%] Built target TriBITS_GenericDriver
Test project /ascldap/users/rabartl/Trilinos.base/BUILDS/GCC-4.8.3/TRIBITS_MPI_DEBUG
    Start 261: TriBITS_CTestDriver_AAOP_ST_ALL_BreakWithSubpackagesALib
    Start 255: TriBITS_CTestDriver_AAO_ST_ALL_BreakWithSubpackagesALib
1/2 Test #261: TriBITS_CTestDriver_AAOP_ST_ALL_BreakWithSubpackagesALib ...   Passed   14.32 sec
2/2 Test #255: TriBITS_CTestDriver_AAO_ST_ALL_BreakWithSubpackagesALib ....   Passed   14.75 sec

100% tests passed, 0 tests failed out of 2

Label Time Summary:
TriBITS    =  29.08 sec (2 tests)

Total Test time (real) =  14.77 sec
```

and these submitted to:

* http://exp.cdash.org/buildSummary.php?buildid=850
* http://exp.cdash.org/index.php?project=TribitsExProj&parentid=845

What is nice is that the latter AAOP build showed the build failures for A.cpp and libpws_a attributed to the WithSubpackages package.  And it did not report any extra build failures for the downstream package's libs or test exectuables.  It seems that CMake/CTest/CDash is smart enough to know when a build failure for downstream lib or executable is due to the failure of an upstream library.  I don't see how this is possible.

Now to keep adding test cases for more failure modes to make sure that all are coverted.

I think I have a good set of test cases now.

[Done]

(6/20/2017)

**6) Finish-up/Clean-up:**

**6.1)** Don't write over `${PROJECT_NAME}_PACAKGES` for the set of packages to directly process.  Instead, set `${PROJECT_NAME}_PACAKGES_TO_DIRECTLY_TEST` and leave `${PROJECT_NAME}_PACAKGES` to be the full list of packages (to be used for other purposes).  (This would allow setting disables on the inner packages as well.) [Done]

**6.2)** Add special logic for failedPackages.txt containing "ALL_PACKAGES" and make sure it enables the full set of packages as the first CI iteration in the follow-up CI iterations. (And add test that checks that case.)

[Done]

(6/21/2017)

**6.3)** Investigate if the all-at-once appraoch with the new CMake/CTest features submitted to the old and new CDash sites can still send out targeted emails based on labels or not (will need to turn on some dummy email addresses and then do some submits to test this out):

This is not going to be that easy to test out.  I think to test this out, I will need to register some dummy email addresses on the CDash sites set up to get emails for certain lables (since TriBITS uses labels for packages still, even after this new CDash feature).

I think I will start by just pinging Kitware about automated testing for CDash emails based on labels.  I did that.

I guess I should to see if I can get emails to get sent out.  Let's do an experiment with the new exp.cdash.org site.  I will register my bartlett.roscoe@gmail.com email address to get CDash emails for one of the TribitsExampleProject packages that has failing tests and then do a submit and see if I can get it to send off emails.

I will hold off on doing this since Zack created:

* https://gitlab.kitware.com/snl/project-1/issues/25

I will wait to test this after that story has completed.

[DEFERRED]

**6.4)** Add automated tests for PBP and AOO for `make dashboard` target:

To do this, I first need to configure a build directory for TribitsExampleProject and then just run the `make dashboard` target.  Therefore, I should be able to use the infrastructure for setting up a build of TribitsExampleProject from the TriBITS/test/core/ExamplesUnitTests/CMakeLists.txt file.

* Test package-by-package 'dashboard' target for different use cases [Done]
* Test all-at-once 'dashboard' target for different use cases [Done]

I added only passing test cases but I feel pretty good about this.

I tested submitting these to CDash with:

```
$ rm -r CMake*

$ ./do-configure -DTriBITS_TRIBITSEXPROJ_SUBMIT_TO=EXP_CDASH

$ ctest -j16 -R "CTestDriver_.+_Dashboard_"
Test project /ascldap/users/rabartl/Trilinos.base/BUILDS/GCC-4.8.3/TRIBITS_MPI_DEBUG
    Start 252: TriBITS_CTestDriver_PBP_Dashboard_ST_ALL_PASS
    Start 253: TriBITS_CTestDriver_PBP_Dashboard_ST_PackagesSubset_PASS
    Start 262: TriBITS_CTestDriver_AAO_Dashboard_ST_ALL_PASS
    Start 263: TriBITS_CTestDriver_AAO_Dashboard_ST_PackagesSubset_PASS
    Start 273: TriBITS_CTestDriver_AAOP_Dashboard_ST_PackagesSubset_PASS
    Start 272: TriBITS_CTestDriver_AAOP_Dashboard_ST_ALL_PASS
1/6 Test #263: TriBITS_CTestDriver_AAO_Dashboard_ST_PackagesSubset_PASS ....   Passed   14.12 sec
2/6 Test #273: TriBITS_CTestDriver_AAOP_Dashboard_ST_PackagesSubset_PASS ...   Passed   14.12 sec
3/6 Test #262: TriBITS_CTestDriver_AAO_Dashboard_ST_ALL_PASS ...............   Passed   16.03 sec
4/6 Test #272: TriBITS_CTestDriver_AAOP_Dashboard_ST_ALL_PASS ..............   Passed   16.83 sec
5/6 Test #253: TriBITS_CTestDriver_PBP_Dashboard_ST_PackagesSubset_PASS ....   Passed   24.43 sec
6/6 Test #252: TriBITS_CTestDriver_PBP_Dashboard_ST_ALL_PASS ...............   Passed   42.77 sec

100% tests passed, 0 tests failed out of 6

Label Time Summary:
TriBITS    = 128.30 sec (6 tests)

Total Test time (real) =  42.78 sec
```

and it reported these tests to:

* http://exp.cdash.org/index.php?project=TribitsExProj&date=2017-06-21

Looks good.

[Done]

**6.5)** Change 'dashboard' target to not do extra configures that are done for that AOO approach (after the tests are in place) [Done]

**6.6)** Add automated tests (and all-at-once functionality) for memcheck and coverage builds (only run if explicitly told to do so at configure time):

Now this is an important test for the all-at-once CDash approach actually.  The question is how will the new CDash site show coverage and valgrind results for the different subprojects.  Will it break them down?  I think it should if possible.  Let's see what happens.

I create the test TriBITS_CTestDriver_PBP_ST_ALL_COVERAGE for the current package-by-package approach and did the CDash submit and it created the page:

* http://exp.cdash.org/index.php?project=TribitsExProj&parentid=885

That page shows the coverage results as well at the bottom.  Very nice.  The front page:

* http://exp.cdash.org/index.php?project=TribitsExProj&date=2017-06-21

shows a link to the coverage results at the bottom under "Coverage" and gives the total percentage of lines covered for the project.  When you click on the line percentage, it takes you to the same page:

* http://exp.cdash.org/index.php?project=TribitsExProj&parentid=885#!#%23Coverage

Very nice.

Now to do valgrind testing for the package-by-package approach.  I added the test `TriBITS_CTestDriver_PBP_ST_ALL_MEMORY` and it submitted to:

* http://exp.cdash.org/index.php?project=TribitsExProj&parentid=890

There is no valgrind warnings so the output is pretty boring.

(6/23/2017)

Now to add support for coverage and memory testing for the all-at-once approach (which I have not implemented yet).

I added the test `TriBITS_CTestDriver_AAOP_ST_ALL_COVERAGE` and did a CDash submit to:

* http://exp.cdash.org/index.php?project=TribitsExProj&parentid=895

and the coverage results submitted separately to:

* http://exp.cdash.org/viewCoverage.php?buildid=895

As you can see, the coverage results are **not** on the same page as the configure, build, and test results like the package-by-package approach and the coverage results are not broken down by subpackage.  This is not good and will need to be fixed.

I also added memory testing and the test `TriBITS_CTestDriver_AAOP_ST_ALL_MEMORY` and submitted to CDash to:

* http://exp.cdash.org/index.php?project=TribitsExProj&parentid=895

and the memory results submitted separately to:

* http://exp.cdash.org/viewDynamicAnalysis.php?buildid=900

As you can see, the memory checking results are **not** on the same page as the configure, build, and tests results like for the package-by-package approach.  That is not good and that will need to be fixed before we upgrade.

[Done]

**6.7)** Add review of coverage and memory testing results to GitLab issue and ask for better views of subproject-by-subproject coverage and memory testing results.

See:

* https://gitlab.kitware.com/snl/project-1/issues/8#note_282638
* https://gitlab.kitware.com/snl/project-1/issues/26

[Done]

**6.8)** Run all of these tests with CMake 3.3.2, 3.6.2 and 'labels_to_subprojects' branches:

Using the `do-configure` script:

```
#!/bin/bash

../../../Trilinos/TriBITS/dev_testing/generic/do-configure-mpi-debug \
-DCTEST_PARALLEL_LEVEL=16 \
-DTriBITS_ENABLE_CONFIGURE_TIMING=ON \
-DTriBITS_CTEST_DRIVER_COVERAGE_TESTS=TRUE \
-DTriBITS_CTEST_DRIVER_MEMORY_TESTS=TRUE \
-DTribitsExProj_USE_NEW_AAO_CTEST_CDASH_FEATURES=TRUE \
"$@"
```

For TriBITS version on the branch:

```
2037f87 "Add all-at-once coverage and memory testing and tests (#183)"
Author: Roscoe A. Bartlett <rabartl@sandia.gov>
Date:   Fri Jun 23 08:54:54 2017 -0600 (2 hours ago)

M       test/ctest_driver/TribitsExampleProject/CMakeLists.txt
M       tribits/ctest_driver/TribitsCTestDriverCoreHelpers.cmake
```

Running tests with CMake 3.6.2:

```
$ which cmake
/home/vera_env/common_tools/cmake-3.6.2/bin/cmake

$ cmake --version
cmake version 3.6.2

$  rm -r CMake*

$ ./do-configure -DTribitsExProj_USE_NEW_AAO_CTEST_CDASH_FEATURES=OFF

$ make

$ time ctest -j16 &> ctest.out

real    0m27.941s
user    3m26.873s
sys     1m24.522s

$ grep -A 50 "tests passed" ctest.out

100% tests passed, 0 tests failed out of 269

Label Time Summary:
TriBITS    = 402.50 sec (269 tests)

Total Test time (real) =  27.93 sec

$ grep "\(_CTestDriver_PBP_\|_CTestDriver_AAO\)" ctest.out | grep "Test "

  3/269 Test #267: TriBITS_CTestDriver_AAO_Dashboard_ST_PackagesSubset_PASS ................................................   Passed    8.42 sec
  5/269 Test #264: TriBITS_CTestDriver_AAO_ST_ALL_COVERAGE .................................................................   Passed    7.91 sec
  6/269 Test #266: TriBITS_CTestDriver_AAO_Dashboard_ST_ALL_PASS ...........................................................   Passed    9.63 sec
  7/269 Test #255: TriBITS_CTestDriver_PBP_Dashboard_ST_PackagesSubset_PASS ................................................   Passed    9.73 sec
  8/269 Test #245: TriBITS_CTestDriver_PBP_ST_ALL_PASS .....................................................................   Passed    9.82 sec
  9/269 Test #250: TriBITS_CTestDriver_PBP_ST_BreakBuildTestPkg ............................................................   Passed   10.02 sec
 10/269 Test #249: TriBITS_CTestDriver_PBP_ST_BreakBuildAllOptionalPkg .....................................................   Passed   10.02 sec
 14/269 Test #252: TriBITS_CTestDriver_PBP_ST_ALL_COVERAGE .................................................................   Passed   11.04 sec
 17/269 Test #265: TriBITS_CTestDriver_AAO_ST_ALL_MEMORY ...................................................................   Passed   14.85 sec
 19/269 Test #251: TriBITS_CTestDriver_PBP_ST_BreakConfigureRequiredPkg ....................................................   Passed    5.09 sec
 20/269 Test #261: TriBITS_CTestDriver_AAO_ST_PackagesSubset_BreakWithSubpackagesALib ......................................   Passed    5.79 sec
 22/269 Test #253: TriBITS_CTestDriver_PBP_ST_ALL_MEMORY ...................................................................   Passed   17.64 sec
 23/269 Test #260: TriBITS_CTestDriver_AAO_ST_ALL_BreakWithSubpackagesALib .................................................   Passed    7.70 sec
 24/269 Test #263: TriBITS_CTestDriver_AAO_ST_ALL_BreakSimpleCxxTestRun ....................................................   Passed    7.80 sec
 25/269 Test #247: TriBITS_CTestDriver_PBP_ST_BreakConfigureOptionalPkg ....................................................   Passed    7.80 sec
 26/269 Test #262: TriBITS_CTestDriver_AAO_ST_ALL_BreakSimpleCxxTestBuild ..................................................   Passed    7.80 sec
 27/269 Test #248: TriBITS_CTestDriver_PBP_ST_BreakBuildLibOptionalPkg .....................................................   Passed    7.81 sec
 30/269 Test #256: TriBITS_CTestDriver_AAO_ST_ALL_PASS .....................................................................   Passed    8.14 sec
 34/269 Test #257: TriBITS_CTestDriver_AAO_ST_PackagesSubset_PASS ..........................................................   Passed    5.03 sec
 67/269 Test #259: TriBITS_CTestDriver_AAO_ST_PackagesSubset_GlobalConfigureFail ...........................................   Passed    0.14 sec
 68/269 Test #246: TriBITS_CTestDriver_PBP_GlobalConfigureFail .............................................................   Passed    0.82 sec
 82/269 Test #258: TriBITS_CTestDriver_AAO_GlobalConfigureFail .............................................................   Passed    0.27 sec
122/269 Test #254: TriBITS_CTestDriver_PBP_Dashboard_ST_ALL_PASS ...........................................................   Passed   10.29 sec
```

Excellent!

Now to test with CMake 3.3.2:

```
$ which cmake
/home/vera_env/common_tools/cmake-3.3.2/bin/cmake

$ cmake --version
cmake version 3.3.2

CMake suite maintained and supported by Kitware (kitware.com/cmake).

$  rm -r CMake*

$ ./do-configure -DTribitsExProj_USE_NEW_AAO_CTEST_CDASH_FEATURES=OFF

$ make

$ time ctest -j16 &> ctest.out

real    0m30.868s
user    3m40.092s
sys     1m25.553s

$ grep -A 50 "tests passed" ctest.out

100% tests passed, 0 tests failed out of 269

Label Time Summary:
TriBITS    = 440.97 sec

Total Test time (real) =  30.60 sec

$ grep "\(_CTestDriver_PBP_\|_CTestDriver_AAO\)" ctest.out | grep "Test "

  4/269 Test #267: TriBITS_CTestDriver_AAO_Dashboard_ST_PackagesSubset_PASS ................................................   Passed    9.52 sec
  5/269 Test #255: TriBITS_CTestDriver_PBP_Dashboard_ST_PackagesSubset_PASS ................................................   Passed   10.97 sec
  6/269 Test #249: TriBITS_CTestDriver_PBP_ST_BreakBuildAllOptionalPkg .....................................................   Passed   11.17 sec
  7/269 Test #266: TriBITS_CTestDriver_AAO_Dashboard_ST_ALL_PASS ...........................................................   Passed   11.27 sec
  8/269 Test #250: TriBITS_CTestDriver_PBP_ST_BreakBuildTestPkg ............................................................   Passed   11.37 sec
 10/269 Test #245: TriBITS_CTestDriver_PBP_ST_ALL_PASS .....................................................................   Passed   11.57 sec
 12/269 Test #252: TriBITS_CTestDriver_PBP_ST_ALL_COVERAGE .................................................................   Passed   12.78 sec
 16/269 Test #265: TriBITS_CTestDriver_AAO_ST_ALL_MEMORY ...................................................................   Passed   14.89 sec
 19/269 Test #261: TriBITS_CTestDriver_AAO_ST_PackagesSubset_BreakWithSubpackagesALib ......................................   Passed    5.12 sec
 20/269 Test #253: TriBITS_CTestDriver_PBP_ST_ALL_MEMORY ...................................................................   Passed   17.16 sec
 21/269 Test #260: TriBITS_CTestDriver_AAO_ST_ALL_BreakWithSubpackagesALib .................................................   Passed    6.96 sec
 23/269 Test #251: TriBITS_CTestDriver_PBP_ST_BreakConfigureRequiredPkg ....................................................   Passed    5.98 sec
 24/269 Test #247: TriBITS_CTestDriver_PBP_ST_BreakConfigureOptionalPkg ....................................................   Passed    7.48 sec
 25/269 Test #263: TriBITS_CTestDriver_AAO_ST_ALL_BreakSimpleCxxTestRun ....................................................   Passed    8.08 sec
 26/269 Test #248: TriBITS_CTestDriver_PBP_ST_BreakBuildLibOptionalPkg .....................................................   Passed    8.39 sec
 27/269 Test #262: TriBITS_CTestDriver_AAO_ST_ALL_BreakSimpleCxxTestBuild ..................................................   Passed    8.39 sec
 29/269 Test #264: TriBITS_CTestDriver_AAO_ST_ALL_COVERAGE .................................................................   Passed    8.59 sec
 34/269 Test #256: TriBITS_CTestDriver_AAO_ST_ALL_PASS .....................................................................   Passed    7.88 sec
 38/269 Test #257: TriBITS_CTestDriver_AAO_ST_PackagesSubset_PASS ..........................................................   Passed    5.72 sec
 65/269 Test #246: TriBITS_CTestDriver_PBP_GlobalConfigureFail .............................................................   Passed    0.64 sec
 70/269 Test #259: TriBITS_CTestDriver_AAO_ST_PackagesSubset_GlobalConfigureFail ...........................................   Passed    0.41 sec
 91/269 Test #258: TriBITS_CTestDriver_AAO_GlobalConfigureFail .............................................................   Passed    0.21 sec
 93/269 Test #254: TriBITS_CTestDriver_PBP_Dashboard_ST_ALL_PASS ...........................................................   Passed   11.41 sec
```

Excellent!

Now to test with CMake 2.8.11:

```
$ which cmake
/home/vera_env/common_tools/cmake-2.8.11/bin/cmake

$ cmake --version
cmake version 2.8.11.2

$  rm -r CMake*

$ ./do-configure -DTribitsExProj_USE_NEW_AAO_CTEST_CDASH_FEATURES=OFF

$ make

$ time ctest -j16 &> ctest.out

real    0m28.597s
user    3m48.940s
sys     0m39.034s

$ grep -A 50 "tests passed" ctest.out

100% tests passed, 0 tests failed out of 246

Label Time Summary:
TriBITS    = 369.34 sec

Total Test time (real) =  28.58 sec

$ grep "\(_CTestDriver_PBP_\|_CTestDriver_AAO\)" ctest.out | grep "Test "

[empty]
```

That is right, I set up the CMakeLists.txt file to not even add these _CTestDriver_ tests for CMake 2.8.11 since I could not get them to work for that version of CMake.  But I really don't care if these tests work for that version of CMake.

And finally, testing with the version of CMake on the 'labels_to_subprojects' branch:

```
$ which cmake
/home/rabartl/install/cmake-labels-to-subprojects-166c3ad-20170617/bin/cmake

$ cmake --version
cmake version 3.9.20170612-g166c3

CMake suite maintained and supported by Kitware (kitware.com/cmake).


$  rm -r CMake*

$ ./do-configure

$ make

$ time ctest -j16 &> ctest.out

real    1m0.653s
user    4m27.992s
sys     1m49.774s

$ grep -A 50 "tests passed" ctest.out

100% tests passed, 0 tests failed out of 281

Label Time Summary:
TriBITS    = 699.13 sec (281 tests)

Total Test time (real) =  60.65 sec

$ grep "\(_CTestDriver_PBP_\|_CTestDriver_AAO\)" ctest.out | grep "Test "

  4/281 Test #279: TriBITS_CTestDriver_AAOP_Dashboard_ST_PackagesSubset_PASS ...............................................   Passed   10.32 sec
  5/281 Test #267: TriBITS_CTestDriver_AAO_Dashboard_ST_PackagesSubset_PASS ................................................   Passed   10.42 sec
  6/281 Test #276: TriBITS_CTestDriver_AAOP_ST_ALL_COVERAGE ................................................................   Passed   11.43 sec
  7/281 Test #278: TriBITS_CTestDriver_AAOP_Dashboard_ST_ALL_PASS ..........................................................   Passed   11.92 sec
  8/281 Test #245: TriBITS_CTestDriver_PBP_ST_ALL_PASS .....................................................................   Passed   11.93 sec
  9/281 Test #255: TriBITS_CTestDriver_PBP_Dashboard_ST_PackagesSubset_PASS ................................................   Passed   12.24 sec
 10/281 Test #250: TriBITS_CTestDriver_PBP_ST_BreakBuildTestPkg ............................................................   Passed   12.33 sec
 11/281 Test #266: TriBITS_CTestDriver_AAO_Dashboard_ST_ALL_PASS ...........................................................   Passed   12.34 sec
 12/281 Test #252: TriBITS_CTestDriver_PBP_ST_ALL_COVERAGE .................................................................   Passed   12.90 sec
 14/281 Test #249: TriBITS_CTestDriver_PBP_ST_BreakBuildAllOptionalPkg .....................................................   Passed   11.81 sec
 17/281 Test #265: TriBITS_CTestDriver_AAO_ST_ALL_MEMORY ...................................................................   Passed   16.43 sec
 18/281 Test #277: TriBITS_CTestDriver_AAOP_ST_ALL_MEMORY ..................................................................   Passed   16.43 sec
 19/281 Test #253: TriBITS_CTestDriver_PBP_ST_ALL_MEMORY ...................................................................   Passed   19.34 sec
 20/281 Test #262: TriBITS_CTestDriver_AAO_ST_ALL_BreakSimpleCxxTestBuild ..................................................   Passed    8.73 sec
 22/281 Test #247: TriBITS_CTestDriver_PBP_ST_BreakConfigureOptionalPkg ....................................................   Passed    9.30 sec
 26/281 Test #248: TriBITS_CTestDriver_PBP_ST_BreakBuildLibOptionalPkg .....................................................   Passed    9.50 sec
 28/281 Test #264: TriBITS_CTestDriver_AAO_ST_ALL_COVERAGE .................................................................   Passed    9.70 sec
 29/281 Test #275: TriBITS_CTestDriver_AAOP_ST_ALL_BreakSimpleCxxTestRun ...................................................   Passed    9.25 sec
 30/281 Test #263: TriBITS_CTestDriver_AAO_ST_ALL_BreakSimpleCxxTestRun ....................................................   Passed    9.55 sec
 31/281 Test #269: TriBITS_CTestDriver_AAOP_ST_PackagesSubset_PASS .........................................................   Passed    8.20 sec
 32/281 Test #251: TriBITS_CTestDriver_PBP_ST_BreakConfigureRequiredPkg ....................................................   Passed    5.40 sec
 33/281 Test #260: TriBITS_CTestDriver_AAO_ST_ALL_BreakWithSubpackagesALib .................................................   Passed   11.41 sec
 34/281 Test #274: TriBITS_CTestDriver_AAOP_ST_ALL_BreakSimpleCxxTestBuild .................................................   Passed   11.52 sec
 35/281 Test #261: TriBITS_CTestDriver_AAO_ST_PackagesSubset_BreakWithSubpackagesALib ......................................   Passed    7.20 sec
 36/281 Test #272: TriBITS_CTestDriver_AAOP_ST_ALL_BreakWithSubpackagesALib ................................................   Passed   10.51 sec
 39/281 Test #273: TriBITS_CTestDriver_AAOP_ST_PackagesSubset_BreakWithSubpackagesALib .....................................   Passed    5.73 sec
 42/281 Test #257: TriBITS_CTestDriver_AAO_ST_PackagesSubset_PASS ..........................................................   Passed    6.03 sec
 47/281 Test #256: TriBITS_CTestDriver_AAO_ST_ALL_PASS .....................................................................   Passed    8.25 sec
 57/281 Test #271: TriBITS_CTestDriver_AAOP_ST_PackagesSubset_GlobalConfigureFail ..........................................   Passed    0.46 sec
 58/281 Test #270: TriBITS_CTestDriver_AAOP_GlobalConfigureFail ............................................................   Passed    0.56 sec
 77/281 Test #246: TriBITS_CTestDriver_PBP_GlobalConfigureFail .............................................................   Passed    0.89 sec
 91/281 Test #259: TriBITS_CTestDriver_AAO_ST_PackagesSubset_GlobalConfigureFail ...........................................   Passed    0.45 sec
103/281 Test #258: TriBITS_CTestDriver_AAO_GlobalConfigureFail .............................................................   Passed    1.01 sec
150/281 Test #254: TriBITS_CTestDriver_PBP_Dashboard_ST_ALL_PASS ...........................................................   Passed   19.51 sec
281/281 Test #268: TriBITS_CTestDriver_AAOP_ST_ALL_PASS ....................................................................   Passed   28.74 sec
```

That is good enough testing for me for the TriBITS tests with different versions of CMake.

**6.9)** Test Trilinos 'make dashboard' with updated TriBITS for AAOP mode with CMake on branch submitting to exp.cdash.org Trilinos project:

Using the `do-configure` script:

```
#!/bin/bash

cmake \
-DTPL_ENABLE_MPI=ON \
-DCMAKE_BUILD_TYPE=RELEASE \
-DTrilinos_ENABLE_DEBUG=ON \
-DBUILD_SHARED_LIBS=ON \
-DTrilinos_ENABLE_SECONDARY_TESTED_CODE=OFF \
-DTrilinos_CONFIGURE_OPTIONS_FILE:STRING=cmake/std/MpiReleaseDebugSharedPtSettings.cmake,cmake/std/BasicCiTestingSettings.cmake \
-DTrilinos_ENABLE_DEBUG_SYMBOLS=ON \
-DTrilinos_ENABLE_EXPLICIT_INSTANTIATION=ON \
-DTrilinos_ENABLE_CONFIGURE_TIMING=ON \
-DTrilinos_ENABLE_TESTS=ON \
-DTrilinos_TRACE_ADD_TEST=ON \
-DCTEST_BUILD_FLAGS=-j16 \
-DCTEST_PARALLEL_LEVEL=16 \
"$@" \
../../../Trilinos
```

Do the configure then `make dashboard`:

```
$ cd ~/Trilinos.base/BUILDS/GCC-4.9.3/MPI_RELEASE_DEBUG_SHARED_PT/

$ rm -r CMake*

$ . ~/Trilinos.base/Trilinos/cmake/load_sems_dev_env.sh 

$ export PATH=/home/rabartl/install/cmake-labels-to-subprojects-166c3ad-20170617/bin:$PATH

$ which cmake
/home/rabartl/install/cmake-labels-to-subprojects-166c3ad-20170617/bin/cmake

$ time ./do-configure \
  -DTrilinos_TRIBITS_DIR:STRING=TriBITS/tribits \
  -DTrilinos_DO_ALL_AT_ONCE=ON \
  -DTrilinos_USE_NEW_AAO_CTEST_CDASH_FEATURES=ON \
  -DCTEST_DROP_SITE="exp.cdash.org" \
  -DCTEST_PROJECT_NAME="Trilinos" \
  -DCTEST_DROP_LOCATION="/submit.php?project=Trilinos" \
  -DTrilinos_ENABLE_ALL_PACKAGES=ON \
  &> configure.out

real    2m11.362s
user    1m23.974s
sys     0m33.280s

$ time make dashboard &> make.dashboard.out

real    3m16.834s
user    34m11.846s
sys     3m34.914s
```

This posted to:

* http://exp.cdash.org/index.php?project=Trilinos&parentid=927

Darn, it aborted with a build failure and it did not report any test failures.  How did that happen?  It reports build failures to Tpetra:

* http://exp.cdash.org/viewBuildError.php?buildid=972

showing:

```
In file included from /ascldap/users/rabartl/Trilinos.base/Trilinos/packages/tpetra/core/src/Tpetra_CrsGraph_SerialWrapperNode.cpp:54:0:
/ascldap/users/rabartl/Trilinos.base/Trilinos/packages/tpetra/core/src/Tpetra_CrsGraph_def.hpp: In instantiation of ‘Tpetra::CrsGraph<LO, GO, N, isClassic>::CrsGraph(const Teuchos::RCP<const Tpetra::Map<LocalOrdinal, GlobalOrdinal, Node> >&, const Teuchos::RCP<const Tpetra::Map<LocalOrdinal, GlobalOrdinal, Node> >&, const local_graph_type&, const Teuchos::RCP<Teuchos::ParameterList>&) [with LocalOrdinal = double; GlobalOrdinal = int; Node = Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial>; bool classic = false; Tpetra::CrsGraph<LO, GO, N, isClassic>::local_graph_type = Kokkos::StaticCrsGraph<double, Kokkos::LayoutLeft, Kokkos::Serial, long unsigned int>; typename Kokkos::ViewTraits<LocalOrdinal*, Kokkos::LayoutLeft, typename Node::device_type::execution_space, void>::size_type = long unsigned int; typename Node::device_type::execution_space = Kokkos::Serial]’:
/ascldap/users/rabartl/Trilinos.base/Trilinos/packages/tpetra/core/src/Tpetra_CrsGraph_SerialWrapperNode.cpp:71:3:   required from here
/ascldap/users/rabartl/Trilinos.base/Trilinos/packages/tpetra/core/src/Tpetra_CrsGraph_def.hpp:639:46: error: no match for call to ‘(Kokkos::StaticCrsGraph<double, Kokkos::LayoutLeft, Kokkos::Serial, long unsigned int>::row_map_type {aka Kokkos::View<const long unsigned int*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace> >}) (LO)’
         if (d_ptrs(rlcid) != d_ptrs(rlcid + 1)) {
                                              ^
In file included from /ascldap/users/rabartl/Trilinos.base/Trilinos/packages/kokkos/core/src/Kokkos_Parallel.hpp:52:0,
                 from /ascldap/users/rabartl/Trilinos.base/Trilinos/packages/kokkos/core/src/Kokkos_Serial.hpp:55,
                 from /ascldap/users/rabartl/Trilinos.base/Trilinos/packages/kokkos/core/src/Kokkos_Core.hpp:53,
                 from /ascldap/users/rabartl/Trilinos.base/Trilinos/packages/teuchos/kokkoscompat/src/KokkosCompat_View.hpp:58,
                 from /ascldap/users/rabartl/Trilinos.base/Trilinos/packages/teuchos/kokkoscompat/src/KokkosCompat_ClassicNodeAPI_Wrapper.hpp:5,
                 from /ascldap/users/rabartl/Trilinos.base/Trilinos/packages/tpetra/classic/NodeAPI/Kokkos_DefaultNode.hpp:48,
                 from /ascldap/users/rabartl/Trilinos.base/Trilinos/packages/tpetra/core/src/Tpetra_CrsGraph_SerialWrapperNode.cpp:45:
/ascldap/users/rabartl/Trilinos.base/Trilinos/packages/kokkos/core/src/Kokkos_View.hpp:454:7: note: candidates are:
 class View : public ViewTraits< DataType , Properties ... > {
       ^
In file included from /ascldap/users/rabartl/Trilinos.base/Trilinos/packages/kokkos/core/src/Kokkos_Parallel.hpp
```

What the heck?  Why is the standard CI build passing but this fails?  The only difference should be the version of TriBITS and CMake being used.  Why would that cause the build of Tpetra to fail?

Let's try just the regular configure, build, and test of Tpetra:

```
$ time ./do-configure \
  -DTrilinos_TRIBITS_DIR:STRING=TriBITS/tribits \
  -DTrilinos_DO_ALL_AT_ONCE=ON -DTrilinos_USE_NEW_AAO_CTEST_CDASH_FEATURES=ON \
  -DCTEST_DROP_SITE="exp.cdash.org" \
  -DCTEST_PROJECT_NAME="Trilinos" \
  -DCTEST_DROP_LOCATION="/submit.php?project=Trilinos" \
  -DTrilinos_ENABLE_Tpetra=ON \
  &> configure.out

real    0m12.051s
user    0m7.558s
sys     0m3.562s

$ time make -j16 &> make.out

real    0m15.122s
user    0m59.952s
sys     0m13.959s
```

Darn, we see build failures too?

Let's try to build from scratch and see what happens:

```
$ which cmake
/home/rabartl/install/cmake-labels-to-subprojects-166c3ad-20170617/bin/cmake

$ rm -r `ls ../MPI_RELEASE_DEBUG_SHARED_PT/ | grep -v do-configure`

$ ls
do-configure

$ time ./do-configure \
  -DTrilinos_TRIBITS_DIR:STRING=TriBITS/tribits \
  -DTrilinos_DO_ALL_AT_ONCE=ON \
  -DTrilinos_USE_NEW_AAO_CTEST_CDASH_FEATURES=ON \
  -DCTEST_DROP_SITE="exp.cdash.org" \
  -DCTEST_PROJECT_NAME="Trilinos" \
  -DCTEST_DROP_LOCATION="/submit.php?project=Trilinos" \
  -DTrilinos_ENABLE_Tpetra=ON \
  &> configure.out

real    0m11.903s
user    0m7.593s
sys     0m3.495s

$ time make -j16 &> make.out

real    1m37.564s
user    10m2.028s
sys     1m8.385s
```

Okay, so it is either the version of CMake or the version of TriBITS as issue here.

We can't downgrade TriBITS and try this out so lets just downgrade CMake and see what happens:

```
$ which cmake
/projects/sems/install/rhel6-x86_64/sems/utility/cmake/3.5.2/bin/cmake

$ cmake --version
cmake version 3.5.2

CMake suite maintained and supported by Kitware (kitware.com/cmake).

$ time ./do-configure \
  -DTrilinos_TRIBITS_DIR:STRING=TriBITS/tribits \
  -DTrilinos_DO_ALL_AT_ONCE=ON \
  -DCTEST_DROP_SITE="exp.cdash.org" \
  -DCTEST_PROJECT_NAME="Trilinos" \
  -DCTEST_DROP_LOCATION="/submit.php?project=Trilinos" \
  -DTrilinos_ENABLE_Tpetra=ON \
  &> configure.out

real    0m11.982s
user    0m7.855s
sys     0m3.287s

$ time make -j16 &> make.out

real    5m41.012s
user    51m52.853s
sys     3m27.552s

$ time ctest -j16 &> ctest.out

real    0m32.924s
user    0m33.067s
sys     0m12.752s

$ grep -A 50 "tests passed" ctest.out 

100% tests passed, 0 tests failed out of 129

Label Time Summary:
Tpetra    = 136.82 sec (129 tests)

Total Test time (real) =  32.92 sec
```

Darn, sure enough, it is the version of CMake breaking the Tpetra build.  Someone at some point will need to dig into this but I don't have time to deal with this right now.

I will just do an all-at-once submit with an older version of CMake without the new features and post to CDash:

```
$ rm -r CMake*


$ time ./do-configure \
  -DTrilinos_TRIBITS_DIR:STRING=TriBITS/tribits \
  -DTrilinos_DO_ALL_AT_ONCE=ON \
  -DCTEST_DROP_SITE="exp.cdash.org" \
  -DCTEST_PROJECT_NAME="Trilinos" \
  -DCTEST_DROP_LOCATION="/submit.php?project=Trilinos" \
  -DTrilinos_ENABLE_ALL_PACKAGES=ON \
  &> configure.out

real    2m21.350s
user    1m50.072s
sys     0m28.707s

$ time make dashboard &> make.dashboard.out

real    38m52.959s
user    379m33.855s
sys     36m33.485s
```

This submitted to:

* http://exp.cdash.org/buildSummary.php?buildid=979

The build all passed but it did not run any tests for some reason.  Why was this the case?

For some reason, `-DTrilinos_ENABLE_TESTS:BOOL=` is getting passed in.  That is not good.  The cause was a bad refactoring that set `-D${PROJECT_NAME}_ENABLE_TESTS:BOOL=${${TRIBITS_PACKAGE}_ENABLE_TESTS}`.  I refactored the code and created a commit to fix this:

```
380ea8e "Fix defect in setting <Project>_ENABLE_TESTS=ON        for all-at-once mode (#183)"
Author: Roscoe A. Bartlett <rabartl@sandia.gov>
Date:   Fri Jun 23 20:42:18 2017 -0600 (4 minutes ago)

M       test/ctest_driver/MockCTestDriver/CMakeLists.txt
M       test/ctest_driver/TribitsExampleProject/CMakeLists.txt
M       tribits/ctest_driver/TribitsCTestDriverCoreHelpers.cmake
```

Running this again:

```
$ time make dashboard &> make.dashboard.out

real    99m5.978s
user    999m44.642s
sys     72m10.674s
```

This submitted to:

* http://exp.cdash.org/index.php?project=Trilinos&filtercount=1&showfilters=1&field1=buildstamp&compare1=61&value1=20170624-0245-Experimental

with build details at:

* http://exp.cdash.org/buildSummary.php?buildid=980

As you can see, that ran 2390 tests.

(6/24/2017)

**6.10)** Change name of `${PROJECT_NAME}_DO_ALL_AT_ONCE` to `${PROJECT_NAME}_CTEST_DO_ALL_AT_ONCE` and `${PROJECT_NAME}_USE_NEW_AAO_CTEST_CDASH_FEATURES` to `${PROJECT_NAME}_CTEST_USE_NEW_AAO_FEATURES`: [Done]

**6.11)** Add documentation for all-at-once approach in TriBITS Build Reference and TriBITS Developers Guide and update tribits/ReleaseNotes.txt [Done]

**6.12)** Copy checkin-test-crf450.sh to checkin-test-crf450-cmake-2.8.11.sh and rename to  checkin-test-crf450-cmake-3.3.2.sh to test with CMake 3.3.2 instead of 2.8.11 (so that it runs all the tests) and enable all the CTestDriver tests that work on that machine.  Then create checkin-test-crf450-do-all-push.sh to run both of the scripts and finally push if they all pass. (The post-push Travis CI testing will test with 2.8.11.):

You have to be very careful about how you write these scripts.  But I think I have it.  I confirmed that the _AAO_ tests are being run by the CMake 3.6.2 builds:

```
$ for dir in MPI_DEBUG SERIAL_RELEASE MPI_DEBUG_CMAKE-3.6.2 SERIAL_RELEASE_CMAKE-3.6.2 ; do echo $dir ; grep AAO $dir/ctest.out | grep "Test " ; done
MPI_DEBUG
 59/246 Test #240: TriBITS_MockCTestDriver_CI_AAO_EnableOnlyModified_ModifiedTeko_FailedAllPackages ........................   Passed    2.13 sec
SERIAL_RELEASE
 53/246 Test #240: TriBITS_MockCTestDriver_CI_AAO_EnableOnlyModified_ModifiedTeko_FailedAllPackages ........................   Passed    3.03 sec
MPI_DEBUG_CMAKE-3.6.2
  5/265 Test #260: TriBITS_CTestDriver_AAO_ST_ALL_BreakSimpleCxxTestBuild ..................................................   Passed    8.75 sec
  6/265 Test #261: TriBITS_CTestDriver_AAO_ST_ALL_BreakSimpleCxxTestRun ....................................................   Passed    8.85 sec
  7/265 Test #254: TriBITS_CTestDriver_AAO_ST_ALL_PASS .....................................................................   Passed    8.85 sec
  9/265 Test #262: TriBITS_CTestDriver_AAO_Dashboard_ST_ALL_PASS ...........................................................   Passed    9.09 sec
 20/265 Test #255: TriBITS_CTestDriver_AAO_ST_PackagesSubset_PASS ..........................................................   Passed    5.51 sec
 22/265 Test #259: TriBITS_CTestDriver_AAO_ST_PackagesSubset_BreakWithSubpackagesALib ......................................   Passed    5.82 sec
 23/265 Test #263: TriBITS_CTestDriver_AAO_Dashboard_ST_PackagesSubset_PASS ................................................   Passed    6.94 sec
 27/265 Test #258: TriBITS_CTestDriver_AAO_ST_ALL_BreakWithSubpackagesALib .................................................   Passed    7.34 sec
 50/265 Test #257: TriBITS_CTestDriver_AAO_ST_PackagesSubset_GlobalConfigureFail ...........................................   Passed    0.41 sec
 82/265 Test #240: TriBITS_MockCTestDriver_CI_AAO_EnableOnlyModified_ModifiedTeko_FailedAllPackages ........................   Passed    0.60 sec
112/265 Test #256: TriBITS_CTestDriver_AAO_GlobalConfigureFail .............................................................   Passed    0.14 sec
SERIAL_RELEASE_CMAKE-3.6.2
 28/265 Test #262: TriBITS_CTestDriver_AAO_Dashboard_ST_ALL_PASS ...........................................................   Passed    8.69 sec
 29/265 Test #254: TriBITS_CTestDriver_AAO_ST_ALL_PASS .....................................................................   Passed   10.40 sec
 30/265 Test #255: TriBITS_CTestDriver_AAO_ST_PackagesSubset_PASS ..........................................................   Passed    7.89 sec
 32/265 Test #260: TriBITS_CTestDriver_AAO_ST_ALL_BreakSimpleCxxTestBuild ..................................................   Passed   10.40 sec
 33/265 Test #263: TriBITS_CTestDriver_AAO_Dashboard_ST_PackagesSubset_PASS ................................................   Passed    6.69 sec
 34/265 Test #259: TriBITS_CTestDriver_AAO_ST_PackagesSubset_BreakWithSubpackagesALib ......................................   Passed    6.79 sec
 35/265 Test #261: TriBITS_CTestDriver_AAO_ST_ALL_BreakSimpleCxxTestRun ....................................................   Passed   10.10 sec
 36/265 Test #258: TriBITS_CTestDriver_AAO_ST_ALL_BreakWithSubpackagesALib .................................................   Passed   10.20 sec
 51/265 Test #256: TriBITS_CTestDriver_AAO_GlobalConfigureFail .............................................................   Passed    0.48 sec
 77/265 Test #240: TriBITS_MockCTestDriver_CI_AAO_EnableOnlyModified_ModifiedTeko_FailedAllPackages ........................   Passed    0.31 sec
 82/265 Test #257: TriBITS_CTestDriver_AAO_ST_PackagesSubset_GlobalConfigureFail ...........................................   Passed    0.31 sec
```

Looks about right.  I can't run the _AAOP_ tests without the CMake/CTest version on 

[Done]

**6.13)** Do a --no-ff merge and push to TriBITS GitHub 'master' using checkin-test-crf450-do-all-push.sh:

With the new checkin-test-crf450-do-all-push.sh I did:

```
$ cd ~/Trilinos.base/Trilinos/TriBITS/
$ git checkout master
$ git merge --no-ff single-config-build-test-submit-183
$ cd ~/Trilinos.base/BUILDS/TRIBITS_CHECKIN/
$ time ./checkin-test-crf450-do-all-push.sh &> checkin-test-crf450-do-all-push.out

real    2m29.681s
user    13m20.379s
sys     3m39.192s
```

[Done]

**6.14)** Snapshot to Trilinos GitHub 'develop' branch:

I updated the TriBITS snapshot and am running the remote pull/test/push script to do full testing and push.

I pushed the commits:

```
commit adf4942d67a2c752f5a08fb4d6d22091998de3ab
Merge: 8b843d0 d40ec9c
Author: Roscoe A. Bartlett <rabartl@sandia.gov>
Date:   Sat Jun 24 13:13:13 2017 -0600

    Merge remote branch 'intermediate-repo/develop' into develop
    
    Build/Test Cases Summary
    Enabled Packages:
    Disabled Packages: PyTrilinos,Claps,TriKota
    Enabled all Packages
    0) MPI_RELEASE_DEBUG_SHARED_PT => passed: passed=2341,notpassed=0 (183.34 min)
    Other local commits for this build/test group: d40ec9c, e669c24

commit d40ec9cf9c80681a08c454f39ab409c5100b8762
Merge: 3e03236 e669c24
Author: Roscoe A. Bartlett <rabartl@sandia.gov>
Date:   Sat Jun 24 13:10:25 2017 -0600

    Merge branch 'tribits_github_snapshot' into develop (TriBITSPub/TriBITS#183)
    
    This provides Trilinos with the ability to do an all-at-once configure, build,
    test, and submit for all the Trilinos packages at once.  Unfortunately,
    results can't be split out for the individual packages on CDash until we
    upgrade to a new version of CMake/CTest and CDash (which are still on branches
    needing more work before considering an upgrade for Trilinos).
    
    But TriBITS should be good to go once CMake/CTest and CDash can be finished
    and upgraded.

commit e669c24edb4f59895ea4351a5e2060a47b8d34a6
Author: Roscoe A. Bartlett <rabartl@sandia.gov>
Date:   Sat Jun 24 13:09:33 2017 -0600

    Automatic snapshot commit from tribits at cfebec9
    
    Origin repo remote tracking branch: 'github/master'
    Origin repo remote repo URL: 'github = git@github.com:TriBITSPub/TriBITS.git'
    
    At commit:
    
    commit 3843b1a2a5fba9dddd41aeba82b4772a0e20e758
    Author:  Roscoe A. Bartlett <rabartl@sandia.gov>
    Date:    Sat Jun 24 10:04:49 2017 -0600
    Summary: Add documentation and release notes for all-at-once mode (#183)

M       cmake/tribits/ReleaseNotes.txt
M       cmake/tribits/ci_support/TribitsDependencies.py
A       cmake/tribits/ci_support/TribitsPackageTestNameUtils.py
A       cmake/tribits/ci_support/get-tribits-packages-from-last-tests-failed.py
M       cmake/tribits/core/package_arch/TribitsGlobalMacros.cmake
M       cmake/tribits/core/package_arch/TribitsPackageMacros.cmake
M       cmake/tribits/core/utils/NinjaMakefileCommon.make
M       cmake/tribits/ctest_driver/TribitsAddDashboardTarget.cmake
M       cmake/tribits/ctest_driver/TribitsCTestDriverCore.cmake
M       cmake/tribits/ctest_driver/TribitsCTestDriverCoreHelpers.cmake
M       cmake/tribits/doc/build_ref/TribitsBuildReferenceBody.rst
M       cmake/tribits/doc/developers_guide/TribitsDevelopersGuide.rst
M       cmake/tribits/examples/TribitsExampleProject/cmake/ctest/general_gcc/ctest_serial_debug.cmake
```

I should have run the remote pull/test/push process from the tracking branch.

[Done]

**Comment by bartlettroscoe** [comment]
Up updated the TriBITS 'master' branch for the all-at-once feature to now match the new CMake/CTest branch:

* https://gitlab.kitware.com/betsy.mcphail/cmake/commits/labels-for-subprojects

for the different way to handle mapping from labels to subprojects.

See:

* https://gitlab.kitware.com/snl/project-1/issues/8

The old branch `single-config-build-test-submit-183` was merged to 'master' and now deleted.   Also, the single branch `single-submit-and-disabled-tests-8-11` was not a good idea and is deleted.  The changes needed for the DISABLED ctest property in #173 where moved to the branch `disabled-test-property-173` so the branch `single-submit-and-disabled-tests-8-11`  has been deleted as well.

I also merged and deleted the new topic branch `all-at-once-ctest-driver-183-2` for this.

**Comment by bartlettroscoe** [comment]
Now that the CMake/CTest `labels-for-subprojects` branch has been merged to the CMake 'master' branch on `git@github.com:kitware/cmake` as of the commit:

```
commit f5be951117711a3965db0f9cd59a2913efacece7
Merge: 37915a6 376dc3e
Author:     Brad King <brad.king@kitware.com>
AuthorDate: Thu Jul 13 11:45:54 2017 +0000
Commit:     Kitware Robot <kwrobot@kitware.com>
CommitDate: Thu Jul 13 07:46:09 2017 -0400

    Merge topic 'labels-for-subprojects'
    
    376dc3eb Help: Add notes for topic 'labels_for_subprojects'
    a70d8e93 Add tests for new directory labels and labels-for-subprojects features
    47b3a57c Display subproject timing summary
    d3859624 Add directory property 'LABELS' and CMAKE_DIRECTORY_LABELS variable
    d08ec4d2 Add CTEST_LABELS_FOR_SUBPROJECTS as a CTest module and script variable
    
    Acked-by: Kitware Robot <kwrobot@kitware.com>
    Merge-request: !1004
```

I think that I can update TriBITS code to automatically set  `${PROJECT_NAME}_CTEST_USE_NEW_AAO_FEATURES=TRUE` when it detects a new-enough version of CMake.  I will have to experiment with this some to see how CMake does this but this should set us up for the next release of CMake/CTest and make one less option to set.   However, I don't think we can set `${PROJECT_NAME}_CTEST_DO_ALL_AT_ONCE=TRUE` when a new-enough version of CMake/CTest is available.  That is because to be useful, an updated version of CDash must be available.  But in any case, it is harmless to set `${PROJECT_NAME}_CTEST_USE_NEW_AAO_FEATURES=TRUE`.

**Comment by bartlettroscoe** [comment]
I got to the bottom of why we were getting all of the TriBITS packages dispalyed on CDash even if only a subset of packages are actually enabled and tested descirbed at:

* https://gitlab.kitware.com/snl/project-1/issues/7#note_294279 

Basically, you have to change the way you set the varaible `CTEST_LABELS_FOR_SUBPROJECTS` for the outer CTest driver.  The updated behavior was committed in d03fddec1de17afcfb04ef9a64fbeee0c58a281d.

Hopefully this will be the last TriBITS change needed to support the new CMake/CTest/CDash all-at-once configure, build, test and submit feature being worked on by Kitware (and soon to be deployed hopefully after some strong manual testing).

----

**DETAILED NOTES:**

I need to get to the bottom of the behavior that I saw reported in:

* https://gitlab.kitware.com/snl/project-1/issues/7#note_294279 

when trying to submit only Teuchos results to CDash and instead it printed a row for every Trilinos package.

I want to examine this behavior with the simple TribitsExProj where I can get very fast turn-around times.

The equivalent will be just enabling SimpleCxx and then running `make dashboard` to see the dashboard results.

I added two new tests to TriBITS that have reproduced the problem.  I have one for the CTEst driver case `TriBITS_CTestDriver_AAOP_ST_SimpleCxx_PASS` and one for the `make dashboard` case `TriBITS_CTestDriver_AAOP_Dashboard_ST_SimpleCxx_PASS`.  In both cases, I only enabled SimpleCxx and I got rows for all four package, even though there were no targets or tests associated with these:

* http://exp.cdash.org/index.php?project=TribitsExProj&parentid=1505
* http://exp.cdash.org/index.php?project=TribitsExProj&parentid=1500

So now I have to figure out why CDash is displaying rows for all of these packages/subprojects.  It is because TriBITS is defining too many subprojects or is it because CDash is always displaying a row for every subproject?

I think the reason might be the implementation of the TriBITS macro:

```
MACRO(TRIBITS_SET_LABELS_TO_SUBPROJECTS_MAPPING)
  IF (${PROJECT_NAME}_CTEST_USE_NEW_AAO_FEATURES)
    SET(CTEST_LABELS_FOR_SUBPROJECTS)
    FOREACH(TRIBITS_PACKAGE ${${PROJECT_NAME}_PACKAGES})
      TRIBITS_IS_PRIMARY_META_PROJECT_PACKAGE(${TRIBITS_PACKAGE}  PACKAGE_IS_PMPP)
      IF (PACKAGE_IS_PMPP)
        LIST(APPEND CTEST_LABELS_FOR_SUBPROJECTS ${TRIBITS_PACKAGE})
       ENDIF()
    ENDFOREACH()
  ENDIF()
ENDMACRO()
```

This macro defines CDash subprojects for every Primary Metra-Project Package (PMPP) TriBITS package defined.  To fix this, we are going to have to change the logic to only define subprojects for packages that are explicitly tested.  The problem is that this function is called from both the inner CMake configure and the outer CTest Driver confiugre as shown by:

```
[rabartl@crf450 tribits (master)]$ find . -name "*.cmake" -exec grep -nHi TRIBITS_SET_LABELS_TO_SUBPROJECTS_MAPPING {} \;
./core/package_arch/TribitsGlobalMacros.cmake:1978:MACRO(TRIBITS_SET_LABELS_TO_SUBPROJECTS_MAPPING)
./core/package_arch/TribitsGlobalMacros.cmake:2010:  TRIBITS_SET_LABELS_TO_SUBPROJECTS_MAPPING()
./ctest_driver/TribitsCTestDriverCoreHelpers.cmake:1177:  TRIBITS_SET_LABELS_TO_SUBPROJECTS_MAPPING()
```

I think we need a different function for the outer CTest driver that just sets the subprojects based on the enabled packages and if we want to expliclty test implicilty enabled packages.

After making the changes in the TriBITS commit d03fddec1de17afcfb04ef9a64fbeee0c58a281d:

```
commit d03fddec1de17afcfb04ef9a64fbeee0c58a281d
Author:     Roscoe A. Bartlett <rabartl@sandia.gov>
AuthorDate: Mon Aug 7 10:22:34 2017 -0600
Commit:     Roscoe A. Bartlett <rabartl@sandia.gov>
CommitDate: Mon Aug 7 10:30:43 2017 -0600

    Fix mapping of labels to subprojects (#183)
    
    I had to create a new macro
    TRIBITS_CTEST_DRIVER_SET_LABELS_TO_SUBPROJECTS_MAPPING() that is only used by
    the outer CTest Driver code.  That macro onlly sets
    CTEST_LABELS_FOR_SUBPROJECTS to the exact set of packages to be explictly
    tested.  That is needed to limit the number of rows that shows up on CDash.
    
    I also simplified the macro TRIBITS_SET_LABELS_TO_SUBPROJECTS_MAPPING() which
    is now only used by the inner CMake configure and only matters when running
    ctest locally.  In this case, there is no reason to limit the set of
    subprojects defined because ctest will only show lines for
    subproejcts/packages that have 1 or more tests defined.
    
    Given that these macros only cat a CMake/CTest var
    CTEST_LABELS_FOR_SUBPROJECTS I took out the inner if statement basd on
    ${PROJECT_NAME}_CTEST_USE_NEW_AAO_FEATURES since this is just setting a var
    which will be ignored by older versions of CMake/CTest.
    
    I added some strong checks to make sure that CTEST_LABELS_FOR_SUBPROJECTS is
    set correctly for each test case and then submitted these cases to CDash to
    confirm that the correct number of rows are showing up.
    
    Build/Test Cases Summary
    Enabled Packages:
    Enabled all Packages
    0) MPI_DEBUG => passed: passed=247,notpassed=0 (0.62 min)
    1) SERIAL_RELEASE => passed: passed=247,notpassed=0 (0.61 min)
    2) MPI_DEBUG_CMAKE-3.6.2 => passed: passed=268,notpassed=0 (0.72 min)
    3) SERIAL_RELEASE_CMAKE-3.6.2 => passed: passed=268,notpassed=0 (0.65 min)

M       test/ctest_driver/TribitsExampleProject/CMakeLists.txt
M       tribits/core/package_arch/TribitsGlobalMacros.cmake
M       tribits/ctest_driver/TribitsCTestDriverCoreHelpers.cmake
```

I submitted the build `TriBITS_CTestDriver_AAOP_ST_SimpleCxx_PASS` to CDash to:

* http://exp.cdash.org/index.php?project=TribitsExProj&parentid=1510

and I get just one row for SimpleCxx.

And I submitted for the builds `TriBITS_CTestDriver_AAOP_ST_PackagesSubset_PASS` and `TriBITS_CTestDriver_AAOP_ST_ALL_PASS` to:

* http://exp.cdash.org/index.php?project=TribitsExProj&parentid=1512
* http://exp.cdash.org/index.php?project=TribitsExProj&parentid=1515

and it looks like we get the correct number of rows shown on CDash.

In and tried this for the builds `TriBITS_CTestDriver_AAOP_Dashboard_ST_PackagesSubset_PASS`, `TriBITS_CTestDriver_AAOP_Dashboard_ST_SimpleCxx_PASS` and `TriBITS_CTestDriver_AAOP_Dashboard_ST_ALL_PASS` submitted to:

* http://exp.cdash.org/index.php?project=TribitsExProj&parentid=1520
* http://exp.cdash.org/index.php?project=TribitsExProj&parentid=1523
* http://exp.cdash.org/index.php?project=TribitsExProj&parentid=1525

and they is displaying the right number of rows.

But when running ctest locally, having all the subprojects defined does not seem to be a problem.  I showed this manually with:

```
$ cd test/ctest_driver/TribitsExampleProject/TriBITS_CTestDriver_AAOP_ST_SimpleCxx_PASS/BUILD/

$ grep Labels DartConfiguration.tcl 
LabelsForSubprojects: SimpleCxx;MixedLang;WithSubpackages;WrapExternal

$ ctest
Test project /ascldap/users/rabartl/Trilinos.base/BUILDS/GCC-4.8.3/TRIBITS_MPI_DEBUG/test/ctest_driver/TribitsExampleProject/TriBITS_CTestDriver_AAOP_ST_SimpleCxx_PASS/BUILD
    Start 1: SimpleCxx_HelloWorldTests
1/2 Test #1: SimpleCxx_HelloWorldTests ........   Passed    0.00 sec
    Start 2: SimpleCxx_HelloWorldProg
2/2 Test #2: SimpleCxx_HelloWorldProg .........   Passed    0.00 sec

100% tests passed, 0 tests failed out of 2

Subproject Time Summary:
SimpleCxx    =   0.00 sec (2 tests)

Total Test time (real) =   0.01 sec
```

**Comment by bartlettroscoe** [comment]
We have been using this for ATDM Trilinos builds for a long time.  Closing as complete.
